pkg/controllers/node/termination/terminator/terminator.go

```diff
@@ -19,46 +19,52 @@ package terminator
 import (
 	"context"
 	"fmt"
+	"time"
 
 	"github.com/samber/lo"
 	v1 "k8s.io/api/core/v1"
 	"k8s.io/apimachinery/pkg/api/equality"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
 	"k8s.io/utils/clock"
+	"knative.dev/pkg/ptr"
 	"sigs.k8s.io/controller-runtime/pkg/client"
 	"sigs.k8s.io/controller-runtime/pkg/log"
 
+	"sigs.k8s.io/karpenter/pkg/events"
 	nodeutil "sigs.k8s.io/karpenter/pkg/utils/node"
 
-	"sigs.k8s.io/karpenter/pkg/apis/v1beta1"
+	terminatorevents "sigs.k8s.io/karpenter/pkg/controllers/node/termination/terminator/events"
 	podutil "sigs.k8s.io/karpenter/pkg/utils/pod"
 )
 
 type Terminator struct {
 	clock         clock.Clock
 	kubeClient    client.Client
 	evictionQueue *Queue
+	recorder      events.Recorder
 }
 
-func NewTerminator(clk clock.Clock, kubeClient client.Client, eq *Queue) *Terminator {
+func NewTerminator(clk clock.Clock, kubeClient client.Client, eq *Queue, recorder events.Recorder) *Terminator {
 	return &Terminator{
 		clock:         clk,
 		kubeClient:    kubeClient,
 		evictionQueue: eq,
+		recorder:      recorder,
 	}
 }
 
-// Taint idempotently adds the karpenter.sh/disruption taint to a node with a NodeClaim
-func (t *Terminator) Taint(ctx context.Context, node *v1.Node) error {
+// Taint idempotently adds an arbitrary taint to a node with a NodeClaim
+func (t *Terminator) Taint(ctx context.Context, node *v1.Node, taint v1.Taint) error {
 	stored := node.DeepCopy()
-	// If the taint already has the karpenter.sh/disruption=disrupting:NoSchedule taint, do nothing.
+	// If the node already has the correct taint (key, value, and effect), do nothing.
 	if _, ok := lo.Find(node.Spec.Taints, func(t v1.Taint) bool {
-		return v1beta1.IsDisruptingTaint(t)
+		return t.MatchTaint(&taint) && t.Value == taint.Value && t.Effect == taint.Effect
 	}); !ok {
-		// If the taint key exists (but with a different value or effect), remove it.
+		// Otherwise, if the taint key exists (but with a different value or effect), remove it.
```

Comment:
You're right, I was thinking this might be a one line change on the condition in `lo.Find`, but it's actually slightly more nuanced. There's a few states:
- A taint with the same key-effect (checked by `MatchTaint`) exists, but a different value. We need to remove our this taint
- A taint with the same key, but different effect exists, we can keep that taint and add our taint. In this case we don't care about the value.
- No taint with the same key exists, we can just add our taint.

This is implemented in a general sense in #1314.

We didn't care about this case when we were just considering the disruption taint, since Karpenter is the only controller which should be managing that. But now it is possible for a user to add an OOS taint with a different effect. Is there a reason someone should do this? I don't think so but it is possible AFAIK.

Reply:
Just to check, we're asking about the desired behavior if the value/effect changes for an existing taint and whether or not this function should update the value/effect on the existing taint?

"Remove" in this function never actually removes the taint, it removes it from the map prior to adding the taint being added back to the map before it's persisted in the k8s api, effectively making it an update.

Reply:
My 2c is that a user adding their version of a taint that Karpenter normally manages (in the karpenter.sh domain) is undefined behavior and it's ok for things to go wrong.

Reply:
Totally agree in the case of any of our specific taints, but this function also manages the OOS taint right? Let's say a user applied a `NoSchedule` version of the OOS taint and we removed it to apply the `NoExecute` version. Technically, I don't think this is correct behavior since theoretically a pod could tolerate the `NoExecute` but not the `NoSchedule`. Does this ever make sense for a user to do? Almost certainly not. But it would be good to have the correct behavior since this is a generalized taint function.

Reply:
Are we fine to keep the generalized version of this taint code around if we aren't applying the OOS taint anymore?

Reply:
I'm good with that 👍